python/ray/spark/__init__.py

```diff
@@ -0,0 +1,607 @@
+    #     the spark job will hang and retry. If spark exceeds the maximum retry count, the
+    #     submitted operation will fail.
+    #  2. Each spark task launches one ray worker node. This design ensures all ray worker nodes
+    #     have the same shape (same cpus / gpus / memory configuration). If ray worker nodes have a
+    #     non-uniform shape, the Ray cluster setup will be non-deterministic and could create
+    #     issues with node sizing.
+    #  3. A ray worker node is started via the `ray start` CLI. In each spark task, a child
+    #     process is started and will execute a `ray start ...` command in blocking mode.
+    #  4. When the ray cluster is shutdown, killing ray worker nodes is implemented by:
+    #     Installing a PR_SET_PDEATHSIG signal for the `ray start ...` child processes
+    #     so that when parent process (pyspark task) is killed, the child processes
+    #     (`ray start ...` processes) will receive a SIGTERM signal, killing it.
+    #     Shutting down the ray cluster is performed by calling `sparkContext.cancelJobGroup`
+    #     to cancel the background spark job, sending a SIGKILL signal to all spark tasks.
+    #     Once the spark tasks are killed, this triggers the sending of a SIGTERM to the child
+    #     processes spawned by the `ray_start ...` process.
+
+    def ray_cluster_job_mapper(_):
+        from pyspark.taskcontext import BarrierTaskContext
+
+        _worker_logger = logging.getLogger("ray.spark.worker")
+
+        context = BarrierTaskContext.get()
+        task_id = context.partitionId()
+
+        task_ip_list = [info.address.split(":")[0] for info in context.getTaskInfos()]
+        task_local_rank = get_spark_task_local_rank(task_id, task_ip_list)
+
+        # NB: If we launch multiple ray worker nodes at the same time,
+        #  it might cause Raylet to have a port conflict, likely due to a race condition.
+        #  A sleep is added here to attempt to avoid resource allocation contention with available
+        #  ports.
+        time.sleep(task_local_rank * 10.0)
+
+        # Ray worker might run on a machine different with the head node, so create the
+        # local log dir and temp dir again.
+        os.makedirs(ray_temp_dir, exist_ok=True)
+        os.makedirs(ray_log_dir, exist_ok=True)
+
+        ray_worker_node_manager_port = get_safe_port()
+        ray_worker_object_manager_port = get_safe_port()
+
+        ray_worker_cmd = [
+            ray_exec_path,
+            "start",
+            f"--temp-dir={ray_temp_dir}",
+            f"--num-cpus={num_spark_task_cpus}",
+            "--block",
+            "--disable-usage-stats",
+            f"--address={ray_head_hostname}:{ray_head_port}",
+            f"--memory={ray_worker_heap_mem_bytes}",
+            f"--object-store-memory={ray_worker_object_store_mem_bytes}",
+            f"--node-manager-port={ray_worker_node_manager_port}",
+            f"--object-manager-port={ray_worker_object_manager_port}",
```

Comment:
TODO: We can remove the 2 option setting, and use ray default ramdom port selection.
Same change in ray head side. @BenWilson2

Reply:
removed in latest commit